src/python/grpcio_tests/tests_aio/unit/server_test.py

```diff
@@ -21,40 +21,140 @@
+        await asyncio.get_event_loop().create_future()
+
+    async def _block_shortly(self, unused_request, unused_context):
+        await asyncio.sleep(test_constants.SHORT_TIMEOUT / 2)
+        return _RESPONSE
+
+    def service(self, handler_details):
+        self._called.set_result(None)
+        if handler_details.method == _SIMPLE_UNARY_UNARY:
+            return grpc.unary_unary_rpc_method_handler(self._unary_unary)
+        if handler_details.method == _BLOCK_FOREVER:
+            return grpc.unary_unary_rpc_method_handler(self._block_forever)
+        if handler_details.method == _BLOCK_SHORTLY:
+            return grpc.unary_unary_rpc_method_handler(self._block_shortly)
+
+    async def wait_for_call(self):
+        await self._called
+
+
+async def _start_test_server():
+    server = aio.server()
+    port = server.add_insecure_port('[::]:0')
+    generic_handler = _GenericHandler()
+    server.add_generic_rpc_handlers((generic_handler,))
+    await server.start()
+    return 'localhost:%d' % port, server, generic_handler
 
 
 class TestServer(AioTestBase):
 
     def test_unary_unary(self):
 
         async def test_unary_unary_body():
-            server = aio.server()
-            port = server.add_insecure_port('[::]:0')
-            server.add_generic_rpc_handlers((GenericHandler(),))
-            await server.start()
+            server_target, _, _ = await _start_test_server()
 
-            async with aio.insecure_channel('localhost:%d' % port) as channel:
-                unary_call = channel.unary_unary(_TEST_METHOD_PATH)
+            async with aio.insecure_channel(server_target) as channel:
+                unary_call = channel.unary_unary(_SIMPLE_UNARY_UNARY)
                 response = await unary_call(_REQUEST)
                 self.assertEqual(response, _RESPONSE)
 
         self.loop.run_until_complete(test_unary_unary_body())
 
+    def test_shutdown(self):
+
+        async def test_shutdown_body():
+            _, server, _ = await _start_test_server()
+            await server.stop(None)
+
+        self.loop.run_until_complete(test_shutdown_body())
+
+    def test_shutdown_after_call(self):
+
+        async def test_shutdown_body():
+            server_target, server, _ = await _start_test_server()
+
+            async with aio.insecure_channel(server_target) as channel:
+                await channel.unary_unary(_SIMPLE_UNARY_UNARY)(_REQUEST)
+
+            await server.stop(None)
+
+        self.loop.run_until_complete(test_shutdown_body())
+
+    def test_graceful_shutdown_success(self):
+
+        async def test_graceful_shutdown_success_body():
+            server_target, server, generic_handler = await _start_test_server()
+
+            channel = aio.insecure_channel(server_target)
+            call_task = self.loop.create_task(
+                channel.unary_unary(_BLOCK_SHORTLY)(_REQUEST))
+            await generic_handler.wait_for_call()
+
+            await server.stop(test_constants.SHORT_TIMEOUT)
```

Comment:
I'm concerned that there may be a race condition here that would cause this test not to actually check what we're hoping to check. Namely, that a calling `stop()` on a server with an ongoing RPC completes successfully.

We decided to instrument the test slightly to detect this condition and turn it into a failure. Then, to take that modified test and to spin up at least a thousand instances to ensure that this does not actually happen.